Refactor `core:text/table`

Feoramund opened the pull request:
_Several kilobytes of lookup tables and one grapheme segmentation algorithm later..._

![unicode-table-alignment](https://github.com/odin-lang/Odin/assets/161657516/03452bed-5c22-4054-8d3c-a6dcdeaf4c05)

Resolves #3432

oskarnp commented:
This is awesome. However, I am just not a fan of breaking the API unless it is neccessary?

oskarnp commented:
I'll review next week.

Feoramund commented:
I changed `colw` to make calculations easier, `set_cell_value_and_alignment` because it was inconsistent with `set_cell_value` and only took a string (not an `any`), `build` to be the place where `Width_Proc` is specified (though this could have a default value if needed), renamed the proc from `ascii` to `plain` per the discussion in the original issue, and `write_text_align` to take a `space` argument instead of `colw`.

The arguments were rotated on `write_text_align` to purposefully break the API because the semantics of the arguments changed but the types did not.

oskarnp commented on core/text/table/table.odin:
```diff
@@ -37,10 +43,19 @@ Table :: struct {
 	tblw: int,          // Width of entire table (including padding, excluding borders)
 }
 
+ascii_width_proc :: proc(str: string) -> int {
+	return len(str)
+}
```
Is this really needed, considering ascii is a subset of utf-8?

Feoramund replied:
Yes. `grapheme_count` has to iterate by UTF-8 codepoints then address several lookup tables by binary search. It's not fast at all. If someone knows in advance that all the text they're handling is ASCII, then it's better to use this proc rather than the Unicode version.

oskarnp commented on core/text/table/table.odin:
```diff
@@ -165,7 +180,7 @@ first_row :: proc(tbl: ^Table) -> int {
 	return header_row(tbl)+1 if tbl.has_header_row else 0
 }
 
-build :: proc(tbl: ^Table) {
+build :: proc(tbl: ^Table, width_proc: Width_Proc) {
```
I think I would prefer a default Width_Proc here. Default to `unicode_width_proc`.

But considering previous comment about ascii being a subset, is Width_Proc needed at all? Any use case for custom width procs you can think of?

oskarnp replied:
Given your comment about `grapheme_count` I change my mind here. Leave it as it is.

oskarnp commented on core/text/table/table.odin:
```diff
@@ -251,8 +265,10 @@ write_html_table :: proc(w: io.Writer, tbl: ^Table) {
 
 write_plain_table :: proc(w: io.Writer, tbl: ^Table) {
+	WIDTH_PROC :: unicode_width_proc
+
 	if tbl.dirty {
-		build(tbl)
+		build(tbl, WIDTH_PROC)
```
If we are to support custom Width_Proc's they should be stored on the `Table` struct at the very least to avoid this situation. Or just remove custom Width_Proc's. Again, is there a use case?

Kelimion replied:
I suppose that if you have a specific font in mind, you could parse its kerning table and glyph info to be more specific. But are you still working in plain text using this package or doing tables in a graphical context?

Feoramund replied:
I did consider if the proc should be on the table or not. I concluded that it shouldn't be on the table, because how the table is printed is separate from what data is stored on the table and HTML tables do not measure width.

oskarnp replied:
Maybe we should consider `if tbl.dirty` at that point to be an error instead to avoid this. 🤔

Feoramund replied:
Do you want `tbl.dirty` done away with?

oskarnp replied:
What would you suggest?

Feoramund replied:
It's a caching situation. The advantage of having `tbl.dirty` around is that it allows re-printing of the table without needing to recalculate column widths. This is assuming there's a situation where a table has been built and is not often changed, but is shown on more than one or two occasions. However, I think if someone wants to really save time, they're better off writing the table once and storing that as a string and avoiding all of the calculations altogether.

In summary, `dirty` avoids recalculating column widths, but in the end, you're still doing all the work to do the printing, which is the bulk of the work being done (all the calls to `io`'s API which potentially has to branch with each call).

My suggestion is to remove `dirty` and offer string caching as a way to save time, if that's a goal for the programmer.

oskarnp replied:
If you don't mind making those changes, then yes I agree that would be a better solution and we should get rid of it.

oskarnp replied:
The user could easily do the string caching themselves passing a custom `io.Writer` so I think we can skip that and just remove `dirty`.

oskarnp commented on core/text/table/table.odin:
```diff
@@ -152,22 +164,116 @@ header :: proc(tbl: ^Table, values: ..any, loc := #caller_location) {
 	tbl.nr_rows += 1
-	for col in 0..<tbl.nr_cols {
+
+	for col in 0 ..< tbl.nr_cols {
```
`a..<b` (without surrounding space) is the preferred Odin style.

Feoramund replied:
I find this style `a ..< b` to be more readable and tend to use it. I did a search to confirm your statement. At the very least, it's more common (~450 instances of `a..<b` versus ~50 of `a ..< b`). I can change it if it's an issue.

oskarnp commented on core/text/table/table.odin:
```diff
@@ -152,22 +164,116 @@ header :: proc(tbl: ^Table, values: ..any, loc := #caller_location) {
+
+	tbl.nr_rows += 1
+
+	for col in 0 ..< tbl.nr_cols {
```
`a..<b` (without surrounding space) is the preferred Odin style.

oskarnp commented on core/text/table/table.odin:
```diff
@@ -152,22 +164,116 @@ header :: proc(tbl: ^Table, values: ..any, loc := #caller_location) {
+
+	tbl.nr_rows += 1
+
+	for col in 0 ..< tbl.nr_cols {
```
`a..<b` (without surrounding space) is the preferred Odin style.

oskarnp commented on core/text/table/table.odin:
```diff
@@ -37,10 +49,19 @@ Table :: struct {
+}
+
+unicode_width_proc :: proc(str: string) -> (width: int) {
+	_, _, width = #force_inline utf8.grapheme_count(str)
```
`utf8.grapheme_count()` calls `utf8.decode_grapheme_clusters()` which allocates implicitly using `context.allocator`. Isn't this a hidden allocation and a memory leak?
All allocators used should be specified up front at `init`.

Feoramund replied:
```
Inputs:
- str: The input string.
- track_graphemes: Whether or not to allocate and return `graphemes` with extra data about each grapheme.
- allocator: (default: context.allocator)

```

```
grapheme_count :: proc(str: string) -> (graphemes, runes, width: int) {
	_, graphemes, runes, width = decode_grapheme_clusters(str, false)
```

```
				if track_graphemes {
					append(&graphemes, Grapheme{
						byte_index,
						rune_count,
						width - last_width,
					})
				}
```

This interpretation is incorrect. In addition, `grapheme_count` is tested under the test suite, and if a memory leak occurs, it would show up in the warnings.

oskarnp replied:
I stand corrected. Ignore my comment.

oskarnp commented on core/text/table/table.odin:
```diff
@@ -137,22 +165,116 @@ header :: proc(tbl: ^Table, values: ..any, loc := #caller_location) {
+
+	tbl.nr_rows += 1
+
+	for col in 0 ..< tbl.nr_cols {
```
`a..<b` (without surrounding space) is the preferred Odin style.

oskarnp commented on core/text/table/table.odin:
```diff
@@ -306,45 +425,58 @@ write_ascii_table :: proc(w: io.Writer, tbl: ^Table) {
+
+		write_row(w, tbl, row, .Center)
+
+		for col in 0 ..< tbl.nr_cols {
```
`a..<b` (without surrounding space) is the preferred Odin style.

I'll stop pointing these out now. Please fix them all :)

oskarnp reviewed (changes requested):
I think `doc.odin` needs some more love. Could you include the example from your PR? And also showcase using different Width_Proc's.

Feoramund commented:
All set.

oskarnp commented:
I can't get the docs.odin to render that unicode table properly. Both GitHub and Sublime Text shows it incorrectly.
Did you paste that properly? Or are they both wrong.

Feoramund commented:
I have confirmed that Alacritty, kitty, WezTerm (with Thai out of alignment for some reason but otherwise fine), Konsole, QTerminal (same Thai issue), LXTerminal, and MATE terminal render it properly. I may not have good all-around Thai fonts installed, but it's otherwise fine on my end.

GitHub's display is browser-dependent, and it certainly doesn't render properly on my end, but I wouldn't expect it to.

I found a Sublime Text editor package for my distribution of Linux and installed it to have a look. With the default settings, all of the Unicode lines are out of alignment (except the actual Younger Futhark runes, amusingly enough). I'm able to get some of them to align if I set certain monospace fonts (Ubuntu Mono), but as this is the first time I've ever used Sublime Text before, I am uncertain if there may be some setting I'd have to toggle to get it to work properly.

It's very strange that it's getting CJK wrong, as they're the most easily predictable in terms of width. Even with a CJK font set, it's still not aligned. It also has issues with rendering the excessive diacritics line, as in, glyph placeholders are appearing.

I tried manually inserting spaces to see if I could align the Chinese row, and even with extra spaces, it's mis-aligned. I.e. there's no amount of space characters I can manually put in that would make the `|` line up. I'm inclined to say this is a rendering issue.

oskarnp commented:
@gingerBill @Kelimion LGTM

Great work, @Feoramund

Kelimion commented:
Thanks as always, @Feoramund.
